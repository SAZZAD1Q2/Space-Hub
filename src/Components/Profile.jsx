import { useSelector } from 'react-redux';

const Profile = () => {
  const allMissions = useSelector((state) => state.missions.missions);
  const myMissions = allMissions.filter((mission) => mission.join === true);
  const rockets = useSelector((state) => state.rockets.rockets);
  const reserveRocket = rockets.filter((rocket) => rocket.reserved === true);
  return (
    <div className="container py-2">
      <div className="row py-5">
        <div className="col-6">
          <h4>My Missions</h4>
          <ul className="list-group ">
            {myMissions.length > 0 ? (myMissions.map((mission) => (
              <li className="list-group-item d-flex justify-content-between align-items-center" key={mission.mission_id}>
                <div>
                  <p className="m-0">{mission.mission_name}</p>
                </div>
              </li>
            )))
              : (<li className="list-group-item">No missions</li>)}
          </ul>
        </div>
        <div className="col-6">
          <h4>My Rockets</h4>
          <ul className="list-group">
            { reserveRocket.length > 0
              ? (reserveRocket.map((rocket) => (
                <li className="list-group-item d-flex justify-content-between align-items-center" key={rocket.id}>
                  <div>
                    <p className="m-0">{rocket.rocket_name}</p>
                  </div>
                </li>
              )))
              : (
                <li className="list-group-item">
                  No rockets
                </li>
              )}
          </ul>
        </div>
      </div>
    </div>
  );
};
export default Profile;
