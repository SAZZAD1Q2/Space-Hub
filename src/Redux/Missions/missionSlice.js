import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const initialState = {
  missions: [],
  isloading: false,
  error: undefined,
};

export const fetchMissions = createAsyncThunk('missions/fetchMission', async (thunkAPI) => {
  try {
    const response = await fetch('https://api.spacexdata.com/v3/missions');
    return response.json();
  } catch (error) {
    return thunkAPI.rejectWithValue('Something went wrong');
  }
});

const missionSlice = createSlice({

  name: 'missions',
  initialState,
  reducers: {
    join: (state, action) => {
      const data = state.missions;
      const newState = data.map((mission) => {
        if (mission.mission_id !== action.payload) return mission;
        return { ...mission, join: true };
      });
      state.missions = newState;
    },
    leave: (state, action) => {
      const data = state.missions;
      const newState = data.map((mission) => {
        if (mission.mission_id !== action.payload) return mission;
        return { ...mission, join: false };
      });
      state.missions = newState;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchMissions.pending, (state) => {
        state.isloading = true;
      })
      .addCase(fetchMissions.fulfilled, (state, action) => {
        state.isloading = false;
        const data = Object.keys(action.payload).map((mission) => {
          const obj = {
            mission_id: action.payload[mission].mission_id,
            mission_name: action.payload[mission].mission_name,
            description: action.payload[mission].description,
            wikipedia: action.payload[mission].wikipedia,
          };
          return obj;
        });
        state.missions = data;
      })
      .addCase(fetchMissions.rejected, (state, action) => {
        state.error = action.payload;
        state.isloading = false;
      });
  },
});

export default missionSlice.reducer;
export const { join, leave } = missionSlice.actions;
