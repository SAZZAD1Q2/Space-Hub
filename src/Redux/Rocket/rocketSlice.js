import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const initialState = {
  rockets: [],
  isloading: false,
  error: null,
};

export const fetchRockets = createAsyncThunk('rockets/fetchRockets', async (ThunkApi) => {
  try {
    const response = await fetch('https://api.spacexdata.com/v4/rockets');
    return response.json();
  } catch (error) {
    return ThunkApi.rejectWithValue('something went wrong');
  }
});
const rocketSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {
    reserve: (state, action) => {
      const data = state.rockets;
      const newState = data.map((rocket) => {
        if (rocket.id !== action.payload) return rocket;
        return { ...rocket, reserved: true };
      });
      state.rockets = newState;
    },
    cancel: (state, action) => {
      const data = state.rockets;
      const newState = data.map((rocket) => {
        if (rocket.id !== action.payload) return rocket;
        return { ...rocket, reserved: false };
      });
      state.rockets = newState;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchRockets.pending, (state) => {
        state.isloading = true;
      })
      .addCase(fetchRockets.fulfilled, (state, action) => {
        state.isloading = false;
        const rocketdata = Object.keys(action.payload).map((rocket) => {
          const data = {
            id: action.payload[rocket].id,
            rocket_name: action.payload[rocket].name,
            description: action.payload[rocket].description,
            flickr_images: action.payload[rocket].flickr_images[0],
            wikipedia: action.payload[rocket].wikipedia,
          };
          return data;
        });
        state.rockets = rocketdata;
      })
      .addCase(fetchRockets.rejected, (state, action) => {
        state.isloading = false;
        state.error = action.payload;
      });
  },
});
export default rocketSlice.reducer;
export const { reserve, cancel } = rocketSlice.actions;
