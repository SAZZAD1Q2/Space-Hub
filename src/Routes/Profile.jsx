const Profile = () => (
  <div className="container py-2">
    <div className="row py-5">
      <div className="col-6">
        <h4>My Missions</h4>
        <ul className="list-group ">
          <li className="list-group-item">No missions</li>
        </ul>
      </div>
      <div className="col-6">
        <h4>My Rockets</h4>
        <ul className="list-group">
          <li className="list-group-item">
            No rockets
          </li>
        </ul>
      </div>
    </div>
  </div>
);
export default Profile;
