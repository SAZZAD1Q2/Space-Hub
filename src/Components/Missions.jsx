import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchMissions, join, leave } from '../Redux/Missions/missionSlice';

const Missions = () => {
  const missionsData = useSelector((state) => state.missions.missions);
  const dispatch = useDispatch();
  useEffect(() => {
    if (missionsData.length === 0) {
      dispatch(fetchMissions());
    }
  });
  return (

    <div className="container my-5">
      <div className="table-responsive table-sm">

        <table className="table table-bordered text-start table-striped">
          <thead>
            <tr>
              <th scope="col" className="col-lg-1 col-md-1 ">Mission</th>
              <th scope="col" className="col-lg-8 col-md-7">Description</th>
              <th scope="col" className="col-lg-1 col-md-1">Status</th>
              <th scope="col" className="col-lg-2 col-md-3">&nbsp;</th>
            </tr>
          </thead>
          <tbody>
            { missionsData
            && missionsData.map((mission) => (
              <tr key={mission.mission_id}>
                <th scope="row">{mission.mission_name}</th>
                <td>{mission.description}</td>
                <td>
                  { mission.join && (
                    <span className="badge text-bg-info text-light">Active Member</span>
                  )}
                  { !mission.join && <span className="badge text-bg-secondary ">Not A Member</span>}
                </td>
                <td className="text-center">
                  {mission.join && (
                    <button type="button" className="btn btn-outline-danger" onClick={() => dispatch(leave(mission.mission_id))}>Leave Mission</button>
                  ) }
                  { !mission.join && (<button type="button" className="btn btn-outline-secondary" onClick={() => dispatch(join(mission.mission_id))}>Join Mission</button>)}
                </td>
              </tr>
            ))}
            { !missionsData && (
            <div>
              <h3>Loading..</h3>
            </div>
            ) }

          </tbody>
        </table>
      </div>
    </div>
  );
};
export default Missions;
