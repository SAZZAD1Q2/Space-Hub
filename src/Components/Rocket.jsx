import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchRockets, reserve, cancel } from '../Redux/Rocket/rocketSlice';

const Rocket = () => {
  const rocketsData = useSelector((state) => state.rockets.rockets);
  const dispatch = useDispatch();

  useEffect(() => {
    if (rocketsData.length === 0) {
      dispatch(fetchRockets());
    }
  }, [dispatch, rocketsData]);

  return (
    <div className="container my-3">
      {rocketsData.length === 0 ? (
        <div>
          <h3>Loading</h3>
        </div>
      ) : (
        rocketsData.map((rocket) => (
          <div className="card my-4" key={rocket.id}>
            <div className="card-body">
              <div className="row">
                <div className="col-md-4">
                  <img src={rocket.flickr_images} className="img-fluid" alt={rocket.rocket_name} />
                </div>
                <div className="col-md-8">
                  <h4 className="card-title">{rocket.rocket_name}</h4>
                  <p className="card-text">
                    {rocket.reserved && (
                      <span className="badge bg-info text-light me-2">Reserved</span>
                    )}
                    {rocket.description}
                  </p>
                  {rocket.reserved && (
                  <button type="button" onClick={() => dispatch(cancel(rocket.id))} className="btn btn-outline-secondary">Cancel Reservation</button>
                  )}
                  {!rocket.reserved && (
                    <button
                      type="button"
                      onClick={() => dispatch(reserve(rocket.id))}
                      className="btn btn-primary"
                    >
                      Reserve Rocket
                    </button>
                  )}
                </div>
              </div>
            </div>
          </div>
        ))
      )}
    </div>
  );
};

export default Rocket;
