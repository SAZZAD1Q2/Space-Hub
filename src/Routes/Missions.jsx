import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchMissions } from '../Redux/Missions/missionSlice';

const Missions = () => {
  let missionsData = useSelector((state) => state.missions.missions);
  const localMissions = JSON.parse(localStorage.getItem('missions')) || [];
  if (localMissions.length > 0) {
    missionsData = localMissions;
  }
  const dispatch = useDispatch();
  useEffect(() => {
    if (missionsData.length === 0) {
      dispatch(fetchMissions());
    }
  });
  return (

    <div className="container my-5">
      <div className="table-responsive table-sm" />
    </div>
  );
};
export default Missions;
