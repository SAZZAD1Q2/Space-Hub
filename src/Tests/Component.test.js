import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../Redux/Store';
import Missions from '../Components/Missions';
import Rocket from '../Components/Rocket';
import Profile from '../Components/Profile';

describe('Component testing', () => {
  test('Mission component should render correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <Missions />
        </Router>
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test('Rocket component should render correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <Rocket />
        </Router>
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test('Profile component should render correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <Profile />
        </Router>
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
