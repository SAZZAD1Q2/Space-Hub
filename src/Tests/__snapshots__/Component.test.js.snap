// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component testing Mission component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="container my-5"
  >
    <div
      class="table-responsive table-sm"
    >
      <table
        class="table table-bordered text-start table-striped"
      >
        <thead>
          <tr>
            <th
              class="col-lg-1 col-md-1 "
              scope="col"
            >
              Mission
            </th>
            <th
              class="col-lg-8 col-md-7"
              scope="col"
            >
              Description
            </th>
            <th
              class="col-lg-1 col-md-1"
              scope="col"
            >
              Status
            </th>
            <th
              class="col-lg-2 col-md-3"
              scope="col"
            >
              Â 
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component testing Profile component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="container py-2"
  >
    <div
      class="row py-5"
    >
      <div
        class="col-6"
      >
        <h4>
          My Missions
        </h4>
        <ul
          class="list-group "
        >
          <li
            class="list-group-item"
          >
            No missions
          </li>
        </ul>
      </div>
      <div
        class="col-6"
      >
        <h4>
          My Rockets
        </h4>
        <ul
          class="list-group"
        >
          <li
            class="list-group-item"
          >
            No rockets
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component testing Rocket component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="container my-3"
  >
    <div>
      <h3>
        Loading
      </h3>
    </div>
  </div>
</DocumentFragment>
`;
